api/tools/src/main/java/org/pmiops/workbench/tools/elastic/ElasticDocument.java

```diff
@@ -0,0 +1,175 @@
+  private enum ElasticType {
+    KEYWORD, INTEGER, DATE, NESTED;
+
+    String lower() {
+      return this.name().toLowerCase();
+    }
+  }
+
+  // An unfortunate manual mapping of column name to BigQuery index in the nested struct query for
+  // domain types. BigQuery does not return a schema for nested types therefore we cannot lookup
+  // result columns by name (only index).
+  private static final Map<String, Integer> BIG_QUERY_COLUMN_INDICES = ImmutableMap.<String, Integer>builder()
+      .put("concept_id", 0)
+      .put("source_concept_id", 1)
+      .put("start_date", 2)
+      .put("age_at_start", 3)
+      .put("visit_type_concept_id", 4)
+      .put("value_as_number", 5)
+      .put("value_as_concept", 6)
+      .build();
+
+  private static final Map<String, Object> NESTED_FOREIGN_SCHEMA = ImmutableMap.of(
+      "type", ElasticType.NESTED.lower(),
+      "properties", ImmutableMap.builder()
+          .put("concept_id", ImmutableMap.of("type", ElasticType.KEYWORD.lower()))
+          .put("source_concept_id", ImmutableMap.of("type", ElasticType.KEYWORD.lower()))
+          // Domain-dependent fields follow (may be unpopulated).
+          .put("start_date", ImmutableMap.of("type", ElasticType.DATE.lower()))
+          .put("age_at_start", ImmutableMap.of("type", ElasticType.INTEGER.lower()))
+          .put("visit_type_concept_id", ImmutableMap.of("type", ElasticType.KEYWORD.lower()))
+          .put("value_as_number", ImmutableMap.of("type", ElasticType.INTEGER.lower()))
+          .put("value_as_concept", ImmutableMap.of("type", ElasticType.KEYWORD.lower()))
+          .build());
+
+  public static final Map<String, Object> PERSON_SCHEMA = ImmutableMap.<String, Object>builder()
+      .put("gender_concept_id", ImmutableMap.of("type", ElasticType.KEYWORD.lower()))
+      .put("condition_concept_ids", ImmutableMap.of("type", ElasticType.KEYWORD.lower()))
+      .put("condition_source_concept_ids", ImmutableMap.of("type", ElasticType.KEYWORD.lower()))
+      .put("conditions", NESTED_FOREIGN_SCHEMA)
+      .build();
+
+  public final String id;
+  public final XContentBuilder source;
+
+  private ElasticDocument(String id, XContentBuilder source) {
+    this.id = id;
+    this.source = source;
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+    ElasticDocument that = (ElasticDocument) o;
+    return id.equals(that.id) &&
+        source.equals(that.source);
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(id, source);
+  }
+
+  /**
+   * Converts a line of BigQuery results JSON to an Elasticsearch document.
+   */
+  public static ElasticDocument fromBigQueryJson(String line) throws IOException {
+    JsonObject doc = new JsonParser().parse(line).getAsJsonObject();
+    String id = doc.remove("_id").getAsString();
```

Comment:
This code doesn't have tests, so liberally commenting important-but-obscure lines is more important. You'd commented on this issue (removing _id_ from the doc) in the other file, but it's warranted here too.

... but then again, it may also be worth adding basic unit tests on this class at some point. Seems reasonably straightforward to do.

Reply:
Yes, I'm thinking about this. Unfortunately to write a meaningful test here, I'll need BigQuery and/or Elastic. It seems there is still no suitable fake for BigQuery locally aside from mocking out its exact behavior (a lot of work, and critical behavior we'd want to test). I'm thinking I might compromise by stubbing the GCS piece with golden data (saved BigQuery output), then running this tool via our "api-local-test" (runs on each master commit / PR), which starts up the API server, then tests. Another alternative would be to write an integration test, but I'd like some more immediate coverage here than "nightly".

Any thoughts?

Reply:
This could be helpful: https://github.com/dadoonet/testcontainers-java-module-elasticsearch